### ISSUE TITLE
Jenkins Jobs should be able to be unit tested

### ISSUE THREAD

**Issue by dlvenable** [opensearch-project/opensearch-build#1450]
### Is your feature request related to a problem? Please describe

This project's current Jenkins unit tests are testing Jenkins libraries. These tests create fake Jenkinsfile jobs which exercise library code. The goal is to ensure the behavior of the library code is consistent.

I would like to add a unit test for the Data Prepper artifacts build. I believe this test shouldn't need to know the details of the library code which it uses. For example, the test should just know that `signArtifacts` was called. It should not care that `signArtifacts`  calls `sh` or uses `withCredentials`. Those are implementation details of `signArtifacts`.

The current problem is that a test cannot successfully `registerAllowedMethod` on library code. This is because `BuildPipelineTest` [registers the library](https://github.com/opensearch-project/opensearch-build/blob/main/tests/jenkins/BuildPipelineTest.groovy#L26-L34).

### Describe the solution you'd like

I propose a simple refactoring to solve this issue:

1. Create a new base class `CommonPipelineTest`. It would not perform any setup. But, it will have the `testPipeline` method.
2. Make `BaseJobPipelineTest` a sub-class of `CommonPipelineTest`
3. (Optional) Create a new base class `BaseJobPipelineTest` which performs the bare minimal setup for testing an actual job. This minimum is mocking the call to load the library.

**Comment by dlvenable** [comment]
Please see PR #1451 for my proposed implementation and a test which uses this new approach.

**Comment by peternied** [comment]
PR https://github.com/opensearch-project/opensearch-build/pull/1451 is the definition of a *unit test*, the unit is the jenkinsfile, and that is all that is tested.  By being explicit in its expectations and very simple to create/maintain.  What @dlvenable created is a great test of the jenkinsfile.

Now to concern of insufficient coverage, how do you catch if the integration between different components is broken with an interface change or a behavioral change? Unit tests (alone) cannot find those defects by design.  *Integration tests* are the answer, these are harder to author because, unlike the unit level, mocking and sample data is typically limited so you can rely on the components to function as intended.

If we were authoring a test for a build workflow the test need a build manifest with legit stuff inside it to be parsed and acted on, there needs to be a fake version of what happens when `build.sh` is called or the test can run in a way to call `build.sh` as if the workflow was running.  This isn't the same thing as the normal environment these tests run, but its a balance for speed and configurability. 

There are other classes of tests called *System testing*, sometimes 'end to end' or 'scenario' tests that fully exercise everything and can be pointed at a different environment to prevent these tests from impacting real users.  These tests would be executed on a jenkins cluster and would point to real s3 buckets, building real build components.  They would fully vet that everything works, but they are a huge effort to create, maintain, and make sure they don't pollute the related resources with unless builds/notifications/whatnot.

Way more detail about testing https://en.wikipedia.org/wiki/Software_testing#Testing_levels.

----

How can we figure out what kind of testing approach to adopt?

I think about adding new processes by picking some standards
- They must be demonstratable, if we don't have an example we shouldn't ask anyone to do it, if someone follows the example it is clear if the followed the intentions.
- They must be maintainable, consider how much code it takes to author a test, and if that tests breaks how clear why that is not correct?
- They must be scalable, if writing one test does not aid you in understand other tests, they will become unmaintainable, if they take to long to execute no one is going to run them.
- They must make it clear the source of the failure, if a test fails but its so convoluted to understand what the nature of the failure was and you have to just rerun it with a debugger its going to stop scaling and stop maintaining.

If we can adopt standards, then we can align the testing approaches to them and use that to weigh relative value and how much effort we want to invest.

@abhinavGupta16 do you have thoughts on what kinds of standards you'd want to measure our tests against?  I see integration related concerns, but how deep should we validate - an example test file might help clarify.

**Comment by abhinavGupta16** [comment]
> PR #1451 is the definition of a _unit test_, the unit is the jenkinsfile, and that is all that is tested. By being explicit in its expectations and very simple to create/maintain. What @dlvenable created is a great test of the jenkinsfile.
> 
> Now to concern of insufficient coverage, how do you catch if the integration between different components is broken with an interface change or a behavioral change? Unit tests (alone) cannot find those defects by design. _Integration tests_ are the answer, these are harder to author because, unlike the unit level, mocking and sample data is typically limited so you can rely on the components to function as intended.
> 
> If we were authoring a test for a build workflow the test need a build manifest with legit stuff inside it to be parsed and acted on, there needs to be a fake version of what happens when `build.sh` is called or the test can run in a way to call `build.sh` as if the workflow was running. This isn't the same thing as the normal environment these tests run, but its a balance for speed and configurability.
> 
> There are other classes of tests called _System testing_, sometimes 'end to end' or 'scenario' tests that fully exercise everything and can be pointed at a different environment to prevent these tests from impacting real users. These tests would be executed on a jenkins cluster and would point to real s3 buckets, building real build components. They would fully vet that everything works, but they are a huge effort to create, maintain, and make sure they don't pollute the related resources with unless builds/notifications/whatnot.
> 
> Way more detail about testing https://en.wikipedia.org/wiki/Software_testing#Testing_levels.
> 
> How can we figure out what kind of testing approach to adopt?
> 
> I think about adding new processes by picking some standards
> 
> * They must be demonstratable, if we don't have an example we shouldn't ask anyone to do it, if someone follows the example it is clear if the followed the intentions.
> * They must be maintainable, consider how much code it takes to author a test, and if that tests breaks how clear why that is not correct?
> * They must be scalable, if writing one test does not aid you in understand other tests, they will become unmaintainable, if they take to long to execute no one is going to run them.
> * They must make it clear the source of the failure, if a test fails but its so convoluted to understand what the nature of the failure was and you have to just rerun it with a debugger its going to stop scaling and stop maintaining.
> 
> If we can adopt standards, then we can align the testing approaches to them and use that to weigh relative value and how much effort we want to invest.
> 
> @abhinavGupta16 do you have thoughts on what kinds of standards you'd want to measure our tests against? I see integration related concerns, but how deep should we validate - an example test file might help clarify.

I think the premise that jenkinsfile is a unit is flawed. `jenkinsfile` is a something that brings in multiple units together to work and perform a job. So yes, we should have integrated regression files for jenkinsjob since the component that it uses are already unit tested.

Since we currently don't have the framework to execute the job, the next best thing is the regression test with regression test files, which extrapolate and lists every command that is executed when the job is run, giving a good overview of what exactly happens when the job is run.

We know that jenkinsfile are critical to test since the potentially can run and execute commands for production. Therefore, not having complete knowledge about how the file is working can be dangerous. The idea behind not mocking the units is that the changes to the library would force us developer to change all the test cases associated to all the existing job ensuring that the the developer is well aware of the blast radius and has tested the jobs affected, given how critical every job is.

An example of such a file would be from the PR https://github.com/opensearch-project/opensearch-build/pull/1437, file - https://github.com/opensearch-project/opensearch-build/blob/main/tests/jenkins/jenkinsjob-regression-files/sign-standalone-artifacts.jenkinsfile.txt which lists down all the commands that are run.

These tests follow suit from the existing framework that we have 
- Which is scalable (since these tests can be easily replicated)
- The source jobs would fail making sure developer is aware of the scope of the changes
- demonstrable - this using the existing framework

Editing the library, I agree would require more changes to the test case and regressions, but I believe that's the idea. It would enable the developer to be aware of the jobs that are affected, so we don't have surprises during releases

In summary, the I think it's imperative we don't mock our libraries because libraries are evolving and changing as they mature. As more and more jobs start to use them, it would be very easy to miss out testing a job what uses it and a possible bad surprise on release day. It definitely is not the best solution for this problem, but probably the best with our current infrastructure.

**Comment by abhinavGupta16** [comment]
A few good reads on when to mock - 

https://www.accenture.com/us-en/blogs/software-engineering-blog/to-mock-or-not-to-mock-is-that-even-a-question
https://chemaclass.es/blog/to-mock-or-not-to-mock/

**Comment by peternied** [comment]
Files:
- Component under test, https://github.com/opensearch-project/opensearch-build/blob/main/jenkins/sign-artifacts/sign-standalone-artifacts.jenkinsfile#L1
- Test configuration, https://github.com/opensearch-project/opensearch-build/blob/main/tests/jenkins/TestSignArtifacts.groovy#L42
- Regression test comparison, https://github.com/opensearch-project/opensearch-build/blob/main/tests/jenkins/jenkinsjob-regression-files/sign-standalone-artifacts.jenkinsfile.txt#L25


I don't share your confidence in that test if we are attempting to manage a potential high blast radius and critical execution, lets consider the purpose of the job, signing artifacts 

https://github.com/opensearch-project/opensearch-build/blob/1c6c4be535925096009b03482def21294828ee19/tests/jenkins/jenkinsjob-regression-files/sign-standalone-artifacts.jenkinsfile.txt#L25
- How do we know that `workspace/sign.sh` will actually generate the expected signature files?
- How do we know that the signature files are even in the right format?

If we really wanted to test this we would execute the signer which makes external service calls and as part of the test we would verify that the signature generated was correct.

----

If we start from the top of the file down we can quickly pick apart the depth of the test

https://github.com/opensearch-project/opensearch-build/blob/1c6c4be535925096009b03482def21294828ee19/jenkins/sign-artifacts/sign-standalone-artifacts.jenkinsfile#L1
- How can we be sure that `jenkins@20211123` is the most up to date library version?

https://github.com/opensearch-project/opensearch-build/blob/1c6c4be535925096009b03482def21294828ee19/jenkins/sign-artifacts/sign-standalone-artifacts.jenkinsfile#L6
- How do we know this agent label is correct and means its an x64 host?

https://github.com/opensearch-project/opensearch-build/blob/1c6c4be535925096009b03482def21294828ee19/jenkins/sign-artifacts/sign-standalone-artifacts.jenkinsfile#L7
- How do we know this docker image exists in the production system and has the proper dependency?

While all three of these potential issues are part of the regression test data, we are *assuming* that they work as expected, but we aren't actually verifying that is the case.  Any of these could break and we have no safety net to catch them.

---

Considering the call outs to not fake out calling sign.sh - do we know what that would take?  The test environment needs to be able to run the python signer client and make a network call to the signer service.

There are many kinds of ideals to strive towards and we can minimize mocks/fakes/dependency injection to keep our verification as legitimate as possible but all of that investment requires a cost and we need to weight it towards our other efforts.

**Comment by peternied** [comment]
[Recapping an offline conversation]

@abhinavGupta16 could you create a PR to update the changes in #1451 to what you'd want?  Great suggestion from @dblock to avoid talking in the abstract and forward the action.

A consideration, we had a legit jenkins workflow bug that was fixed in this PR https://github.com/opensearch-project/opensearch-build/pull/1467 - `wget` wasn't on the image.  I wish we there was a way we could have detected this in advance.

**Comment by abhinavGupta16** [comment]
> [Recapping an offline conversation]
> 
> @abhinavGupta16 could you create a PR to update the changes in #1451 to what you'd want? Great suggestion from @dblock to avoid talking in the abstract and forward the action.
> 
> A consideration, we had a legit jenkins workflow bug that was fixed in this PR #1467 - `wget` wasn't on the image. I wish we there was a way we could have detected this in advance.

Ideally, we should not use wget or curl to download the artifacts since we know the s3 bucket for the artifacts.
